### DISCUSSION TITLE
Listening for events on presence channel: Echo.join(...)[event_name] is not a function

### DISCUSSION THREAD

**Discussion by redbastie** [livewire/livewire#3071] [7. Bugs (archived)]
Trying to listen for an event on a presence channel and get this error:

> LaravelEcho.js:38 Uncaught TypeError: Echo.join(...)[event_name] is not a function

Here is my code:

    protected $listeners = [
        'echo-presence:lobby,MatchMade' => 'matchMade',
    ];

If this is incorrect, how do I properly listen for events on a presence channel via the listeners property?

**Comment by joshhanley** [comment]
I've investigated this issue and will try to summarise the main problem here as it has been raised a few times (listed for reference) #1845, #1075, #1074. 

I also found an article where they had to work around this issue using Alpine to implement the listen events, which doesn't feel very clean to me: https://medium.com/@zachvv11/real-time-chat-room-with-larasocket-tailwind-alpine-livewire-and-laravel-406a8c5e680d 

## Problem

Laravel Echo currently has 4 main methods when interacting with a presence channel:
- here
- joining
- leaving
- listen

For the first three methods, the signature is only one parameter which is a closure, which we can currently use in Livewire using the listeners property:
```php
protected $listeners = [
        'echo-presence:demo,here' => 'here',
        'echo-presence:demo,joining' => 'joining',
        'echo-presence:demo,leaving' => 'leaving',
    ];
```

But the method signature for `listen` has two parameters, the first is the event class name, and the second is a closure. So currently, you cannot do the below with a presence channel as listen is not specified anywhere:
```php
protected $listeners = [
        'echo-presence:demo,SomeCustomEventClass' => 'newMethod',
];
```
Which is what is then causing this error as (using the above example) it is trying to call:
```js
Echo.join(...)['SomeCustomEventClass']
```
> LaravelEcho.js:38 Uncaught TypeError: Echo.join(...)[event_name] is not a function

What it should call is:
```js
Echo.join(...)['listen']('SomeCustomEventClass', (e) => {...})
```

## Proposed Solutions

The echo-presence signature should be changed to be able to specify when is it both the listen method and what event it should listen for.

One option is to append the event name to the end of the signature after appending listen like:
```php
'echo-presence:demo,listen,SomeCustomEventClass' => ...
```

The problem with that is the current signature special syntax is `'echo:{channel},{event}' => '{method}'`
Which then would change just for this action and look like `'echo:{channel},{action},{event}' => '{method}'` which is inconsistent.

Or another potential option is to actually have a new "channel" defined specifically for presenceListen, such as:
```php
'echo-presenceListen:demo,SomeCustomEventClass' => ...
```

This would also be easier to implement as you don't need to change the way the event parts are constructed, instead just add another conditional for the new channel type.

@calebporzio I'm happy to PR a fix for this, but would like some feedback as to what would be preferable.

Hope this helps!

**Comment by redbastie** [comment]
Or we can just detect the default presence listener names in the codebase and use the existing syntax for those, otherwise use the `listen` syntax. This is the simplest and cleanest solution.

**Comment by joshhanley** [comment]
@redbastie good idea, that's much cleaner!

**Comment by davidpiesse** [comment]
Hey all,
When I wrote the Echo code this was the biggest issue I hit.

It seems this code is wrong.
```js
  } else if (channel_type == 'presence') {
      Echo.join(channel)[event_name](e => {
          store.emit(event, e)
      })
  }
```
In the Laravel docs it has
```js
Echo.join(`chat.${roomId}`)
    .here(...)
    .joining(...)
    .leaving(...)
    .listen('NewMessage', (e) => {
        //
    });
```
So the code cannot currently understand the listen part of a presence channel...
IIRC I tried to get this working but never did...

**Comment by Wulfheart** [comment]
Any update on this?

**Comment by jakebooy2** [comment]
Will this issue be fixed? I want to use presence channels in my application but without the Livewire integration it's a lot more work.

**Comment by alexiovay** [comment]
Still getting the error

**Comment by sidharthmenon** [comment]
The simplest work around I am using is. Not sure it is the correct way but it works.

```
protected $listeners = [
        'echo:presence-demo,SomeCustomEventClass' => 'newMethod',
];
```

**Reply by mrpritchett** [reply]
How is this a workaround? Simply using pubic channels? That's not a workaround at all???

**Comment by mrpritchett** [comment]
This is still an issue. Can a maintainer elaborate why this was moved to a discussion from an issue? It's a bug. The implementation seems to be incorrect or broken.

**Reply by davidpiesse** [reply]
All issues were moved to discussions a while back.

It appears we need some consensus on an option to proffer to the Livewire team.

I prefer the `echo-presenceListen` option of those @joshhanley suggested. 
I'm happy to help implement alongside the core team.

We would also need an extra bit of the documentation that highlights this edge case.

**Reply by joshhanley** [reply]
@mrpritchett yeah @davidpiesse is correct, we're no longer using the issues tab in GitHub as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/livewire/livewire/discussions/3147

In regards to this problem, yeah I was waiting on some feedback before implementing a fix.

I'll see if I can get Caleb to have a look and see which he prefers.

**Reply by calebporzio** [reply]
I don't like `echo-presenceListen` for consistency reasons and also aesthetics (mixing casings)

I think I like this syntax the best:

`protected $listeners = [
        'echo-presence:demo,SomeCustomEventClass' => 'newMethod',
];`

It would need to be documented that this is how you use "listen", but the implementation would be pretty straightforward:
- check if the second param (after the `,`) is not in the list of accepted methods (here, joining, etc...) and default to listen using this param as the event name.

My next favorite would be: `echo-presence:demo,listen:SomeCustomEventClass` or something else similar to `:` - and parse for this in JavaScript.

Thanks for the input everyone!

**Reply by mrpritchett** [reply]
I appreciate everyone weighing in. Y'all are doing awesome work and it's appreciated it. 👍🏻

**Reply by davidpiesse** [reply]
I will give it a go today to get a failing test and hopefully a sufficient PR.

**Reply by joshhanley** [reply]
Yep I think the first option @calebporzio said checking for the accepted methods will work. And yeah agreed much cleaner.

@davidpiesse thanks! Let me know if you have any issues or don't get a chance and I will PR it 🙂

**Reply by davidpiesse** [reply]
I have been out of the game too long it seems @joshhanley !

The integrations / 3rd party areas have no longer built in tests within Livewire.
Do you have an idea of how we might be able to use Dusk for Laravel Echo integration tests? 
I am already barking down the 'second app' path to try and get some manual testing -- and I am not kewl with just roughshod changes without them.

**Reply by joshhanley** [reply]
@davidpiesse hmm I don't think there are any echo tests due to the reliance on a websockets server.

I have a test app setup from a while back with Echo and Pusher running, so once you PR it, I'll pull it and test it as well.

**Reply by davidpiesse** [reply]
PR #4772 should fix this :)
